# CVE-2019-13115

## About
* <https://nvd.nist.gov/vuln/detail/CVE-2019-13115>

## PoC / Exploit
* <https://github.com/kevinbackhouse/security-lab/tree/master/SecurityExploits/libssh2/out_of_bounds_read_kex_CVE-2019-13115>

## Details

The source location of the vulnerability is due to an incorrectly implemented bounds checking function:
 
 ```c
 int _libssh2_check_length(struct string_buf *buf, size_t len) {
     return ((int) (buf->dataptr - buf->data) <= (int) (buf->len - len)) ? 1 : 0;
 }
 ```

The PoC works by making `len` greater than `buf->len + 0x80000000` so that the calculation of `(int)(buf->len - len)` overflows and becomes a very large positive number, thereby bypassing the bounds check and causing libssh2 to crash with a segmentation fault. 


## Environment used

* Ubuntu 18.04.4 LTS
* libssh2 1.82


## Reproduction
* Server
    - Docker:
        ```shell script
        sudo docker build -t cve-2019-13115 . 
        sudo docker run -p 8022:8022 -it cve-2019-13115
        ```

    - Script:
        ```shell script
        bash installation.sh
        bash run.sh
        ```      

* Client
    ```shell script
    bash exploit.sh
    ```

<br>
The result should be a message with <i>"Segmentation fault"</i> in the client and service down in the server.
